plugins/jabra-gnp/fu-jabra-gnp-device.c

```diff
@@ -0,0 +1,707 @@
+
+	if (!fu_device_retry_full(FU_DEVICE(self),
+				  fu_jabra_gnp_device_tx_cb,
+				  FU_JABRA_GNP_MAX_RETRIES,
+				  FU_JABRA_GNP_RETRY_DELAY,
+				  (gpointer)&tx_data,
+				  error))
+		return FALSE;
+	if (!fu_device_retry_full(FU_DEVICE(self),
+				  fu_jabra_gnp_device_rx_with_sequence_cb,
+				  FU_JABRA_GNP_MAX_RETRIES,
+				  FU_JABRA_GNP_RETRY_DELAY,
+				  (gpointer)&rx_data,
+				  error))
+		return FALSE;
+	if (rx_data.rxbuf[5] != 0xFF) {
+		g_set_error(error,
+			    FWUPD_ERROR,
+			    FWUPD_ERROR_INTERNAL,
+			    "internal error: expected 0xFF, got 0x%02x 0x%02x",
+			    rx_data.rxbuf[5],
+			    rx_data.rxbuf[6]);
+		return FALSE;
+	}
+	return TRUE;
+}
+
+static gboolean
+fu_jabra_gnp_device_write_chunk(FuJabraGnpDevice *self,
+				guint32 chunk_number,
+				const guint8 *buf,
+				guint32 bufsz,
+				GError **error)
+{
+	guint8 write_length = 0x00 + bufsz + 10;
+	guint8 txbuf[FU_JABRA_GNP_BUF_SIZE] =
+	    {FU_JABRA_GNP_IFACE, 0x08, 0x00, 0x00, write_length, 0x0F, 0x1A};
+	FuJabraGnpTxData tx_data = {.txbuf = txbuf, .timeout = FU_JABRA_GNP_STANDARD_SEND_TIMEOUT};
+	fu_memwrite_uint16(txbuf + 7, chunk_number, G_LITTLE_ENDIAN);
+	fu_memwrite_uint16(txbuf + 9, bufsz, G_LITTLE_ENDIAN);
+	if (!fu_memcpy_safe(txbuf, sizeof(txbuf), 11, buf, bufsz, 0x0, bufsz, error))
+		return FALSE;
+	return fu_device_retry_full(FU_DEVICE(self),
+				    fu_jabra_gnp_device_tx_cb,
+				    FU_JABRA_GNP_MAX_RETRIES,
+				    FU_JABRA_GNP_RETRY_DELAY,
+				    (gpointer)&tx_data,
+				    error);
+}
+
+static gboolean
+fu_jabra_gnp_device_write_chunks(FuJabraGnpDevice *self,
+				 GPtrArray *chunks,
+				 FuProgress *progress,
+				 GError **error)
+{
+	guint32 preload_count = 100;
+	guint8 rxbuf[FU_JABRA_GNP_BUF_SIZE] = {0x00};
+	const guint8 match_buf[FU_JABRA_GNP_BUF_SIZE] =
+	    {FU_JABRA_GNP_IFACE, 0x08, 0x00, 0x00, 0x06, 0x0F, 0x1B};
+	FuJabraGnpRxData rx_data = {.rxbuf = rxbuf, .timeout = FU_JABRA_GNP_LONG_RECEIVE_TIMEOUT};
+
+	/* progress */
+	fu_progress_set_id(progress, G_STRLOC);
+	fu_progress_set_steps(progress, chunks->len);
+	for (guint chunk_number = 0; chunk_number < chunks->len; chunk_number++) {
+		FuChunk *chk = g_ptr_array_index(chunks, chunk_number);
+		if (!fu_jabra_gnp_device_write_chunk(self,
+						     chunk_number,
+						     fu_chunk_get_data(chk),
+						     fu_chunk_get_data_sz(chk),
+						     error))
+			return FALSE;
+		if (((chunk_number % preload_count) == 0) || chunk_number == chunks->len - 1) {
+			if (!fu_device_retry_full(FU_DEVICE(self),
+						  fu_jabra_gnp_device_rx_cb,
+						  FU_JABRA_GNP_MAX_RETRIES,
+						  FU_JABRA_GNP_RETRY_DELAY,
+						  (gpointer)&rx_data,
+						  error))
+				return FALSE;
+			if (rx_data.rxbuf[5] != match_buf[5] || rx_data.rxbuf[6] != match_buf[6]) {
+				g_set_error(error,
+					    FWUPD_ERROR,
+					    FWUPD_ERROR_INTERNAL,
+					    "internal error, buf did not match");
+				return FALSE;
+			}
+			if (0) {
```

Comment:
Forgot?

Reply:
Not exactly. That is an extra check we can use to validate the firmware bytes being sent to the device, but it may be a bit too restrictive, so I left it in for review, but disabled the check until we start testing the code internally.